kube/src/service/auth.rs

```diff
@@ -131,3 +131,61 @@ where
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::matches;
+
+    use futures::pin_mut;
+    use http::{HeaderValue, Request, Response};
+    use hyper::Body;
+    use tokio_test::assert_ready_ok;
+    use tower_test::mock;
+
+    use crate::{error::ConfigError, Error};
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn valid_token() {
+        const TOKEN: &str = "Bearer test";
+        let auth = Authentication::Token(TOKEN.into());
+        let (mut service, handle) = mock::spawn_layer(AuthLayer::new(auth));
+
+        let spawned = tokio::spawn(async move {
+            // Receive the requests and respond
+            pin_mut!(handle);
+            let (request, send) = handle.next_request().await.expect("service not called");
+            assert_eq!(
+                request.headers().get(AUTHORIZATION).unwrap(),
+                HeaderValue::from_static(TOKEN)
+            );
+            send.send_response(Response::builder().body(Body::empty()).unwrap());
+        });
+
+        assert_ready_ok!(service.poll_ready());
+        service
+            .call(Request::builder().uri("/").body(Body::empty()).unwrap())
+            .await
+            .unwrap();
+        spawned.await.unwrap();
+    }
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn invalid_token() {
+        const TOKEN: &str = "\n";
+        let auth = Authentication::Token(TOKEN.into());
+        let (mut service, _handle) =
+            mock::spawn_layer::<Request<Body>, Response<Body>, _>(AuthLayer::new(auth));
+        let err = service
+            .call(Request::builder().uri("/").body(Body::empty()).unwrap())
+            .await
+            .unwrap_err();
+
+        assert!(err.is::<Error>());
+        assert!(matches!(
+            *err.downcast::<Error>().unwrap(),
+            Error::Kubeconfig(ConfigError::InvalidBearerToken(_))
```

Comment:
What is `err` before the downcast here? I thought our service would already return our error type.

Reply:
`kube::Service` uses [`tower::BoxError`](https://docs.rs/tower/0.4.4/tower/type.BoxError.html). I thought it was necessary to make it composable. So `AuthLayer` does too. But I might be missing something.

`Client::send` looks like this
```rust
    async fn send(&self, request: Request<Body>) -> Result<Response<Body>> {
        let mut svc = self.inner.clone();
        let res = svc
            .ready_and()
            .await
            .map_err(Error::Service)?
            .call(request)
            .await
            .map_err(|err| {
                if err.is::<Error>() {
                    // Error decorating request
                    *err.downcast::<Error>().expect("kube::Error")
                } else if err.is::<hyper::Error>() {
                    // Error requesting
                    Error::HyperError(*err.downcast::<hyper::Error>().expect("hyper::Error"))
                } else {
                    // Errors from other middlewares
                    Error::Service(err)
                }
            })?;
        Ok(res)
    }
```

Reply:
Yeah, make sense. Had a quick go and hit my head against a lot of errors trying to do otherwise.

Since `Client::send` abstracts away this rougher error interface, then it's no problem. `kube::Api` errors usage is the same.

I guess the idea is that if anyone in the future wanting to control their own client, those are the only ones who would have to rewrite `kube::Client` and deal with this type of error matching.